File: backend/internal/telegram/real_client.go
package telegram

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

// RealClient реальная реализация Telegram клиента
type RealClient struct {
	httpClient *http.Client
}

// NewRealClient создает новый реальный Telegram клиент
func NewRealClient() *RealClient {
	return &RealClient{
		httpClient: &http.Client{},
	}
}

// SendMessage отправляет сообщение через Telegram Bot API
func (r *RealClient) SendMessage(token, chatID, message string) error {
	url := fmt.Sprintf("https://api.telegram.org/bot%s/sendMessage", token)

	payload := map[string]string{
		"chat_id": chatID,
		"text":    message,
	}

	jsonData, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("failed to marshal payload: %w", err)
	}

	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := r.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("telegram API error: status=%d, body=%s", resp.StatusCode, string(body))
	}

	return nil
}

File: backend2/src/Telegram/RealClient.php
<?php

namespace App\Telegram;

use Psr\Log\LoggerInterface;

class RealClient implements ClientInterface
{
    public function __construct(
        private readonly LoggerInterface $logger,
        private readonly \Symfony\Contracts\HttpClient\HttpClientInterface $httpClient
    ) {}

    public function sendMessage(string $token, string $chatId, string $message): bool
    {
        $url = sprintf('https://api.telegram.org/bot%s/sendMessage', $token);

        $payload = [
            'chat_id' => $chatId,
            'text' => $message,
        ];

        try {
            $response = $this->httpClient->request('POST', $url, [
                'json' => $payload,
                'headers' => [
                    'Content-Type' => 'application/json',
                ],
            ]);

            $statusCode = $response->getStatusCode();
            $responseData = $response->toArray(false);

            if ($statusCode !== 200) {
                $this->logger->error('Telegram API error', [
                    'status' => $statusCode,
                    'response' => $responseData,
                    'token' => substr($token, 0, 10) . '...',
                    'chat_id' => $chatId,
                ]);
                return false;
            }

            return true;
        } catch (\Exception $e) {
            $this->logger->error('Failed to send Telegram message', [
                'error' => $e->getMessage(),
                'token' => substr($token, 0, 10) . '...',
                'chat_id' => $chatId,
            ]);
            return false;
        }
    }
}
